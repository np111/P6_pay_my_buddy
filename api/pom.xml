<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>com.paymybuddy</groupId>
    <artifactId>paymybuddy</artifactId>
    <version>1.0-SNAPSHOT</version>
    <relativePath>../pom.xml</relativePath>
  </parent>
  <artifactId>paymybuddy-api</artifactId>
  <packaging>jar</packaging>

  <dependencies>
    <!-- Validation API (JSR 380) -->
    <dependency>
      <groupId>javax.validation</groupId>
      <artifactId>validation-api</artifactId>
      <scope>provided</scope>
    </dependency>

    <!-- Jackson Databind API -->
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <scope>provided</scope>
    </dependency>
  </dependencies>
</project>
