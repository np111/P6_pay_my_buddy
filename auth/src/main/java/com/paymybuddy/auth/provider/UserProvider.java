package com.paymybuddy.auth.provider;

import com.paymybuddy.api.model.user.User;
import org.springframework.lang.Nullable;

public interface UserProvider {
    @Nullable
    User getUserById(long userId);

    @Nullable
    User getUserByEmail(String email);

    void updateEncodedPassword(User user, String encodedPassword);
}
