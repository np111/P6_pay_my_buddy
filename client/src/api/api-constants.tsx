export interface CurrencyProps {
    decimals: number;
    prefix: string;
    suffix: string;
}

export const currencies: { [currency: string]: CurrencyProps } = {
    USD: {decimals: 2, prefix: '$', suffix: ' USD'},
    EUR: {decimals: 2, prefix: '', suffix: '€'},
    JPY: {decimals: 0, prefix: '¥', suffix: ''},
    GBP: {decimals: 2, prefix: '£', suffix: ''},
    CHF: {decimals: 2, prefix: '', suffix: ' CHF'},
    CAD: {decimals: 2, prefix: '$', suffix: ' CAD'},
    AUD: {decimals: 2, prefix: '$', suffix: ' AUD'},
    HKD: {decimals: 2, prefix: 'HK$ ', suffix: ''},
};

export const currenciesOptions = Object.keys(currencies).sort().map((currency) => ({value: currency}));
