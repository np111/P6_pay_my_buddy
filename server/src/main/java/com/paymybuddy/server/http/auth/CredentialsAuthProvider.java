package com.paymybuddy.server.http.auth;

import com.paymybuddy.server.service.AuthService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Component
@Scope("singleton")
public class CredentialsAuthProvider implements AuthenticationProvider {
    private final AuthService authService;

    @Override
    public AuthToken authenticate(Authentication authentication) throws AuthenticationException {
        UsernamePasswordAuthenticationToken credentials = (UsernamePasswordAuthenticationToken) authentication;
        return authService.login(determineEmail(credentials), determinePassword(credentials));
    }

    private String determineEmail(Authentication authentication) {
        return authentication.getPrincipal() == null ? "NONE_PROVIDED" : authentication.getName();
    }

    private String determinePassword(Authentication authentication) {
        return authentication.getCredentials().toString();
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return UsernamePasswordAuthenticationToken.class.isAssignableFrom(authentication);
    }
}
