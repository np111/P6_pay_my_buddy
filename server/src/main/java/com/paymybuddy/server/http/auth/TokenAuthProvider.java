package com.paymybuddy.server.http.auth;

import com.paymybuddy.server.service.AuthService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Component
@Scope("singleton")
public class TokenAuthProvider implements AuthenticationProvider {
    private final AuthService authService;

    @Override
    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        return authService.authenticateAuthToken((AuthToken) authentication);
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return AuthToken.class.isAssignableFrom(authentication);
    }
}
