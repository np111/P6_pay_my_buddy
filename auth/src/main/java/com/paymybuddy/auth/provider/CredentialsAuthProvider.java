package com.paymybuddy.auth.provider;

import com.paymybuddy.auth.AuthService;
import com.paymybuddy.auth.AuthToken;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Component
@Scope("singleton")
public class CredentialsAuthProvider implements AuthenticationProvider {
    private final AuthService authService;

    @Override
    public AuthToken authenticate(Authentication authentication) throws AuthenticationException {
        UsernamePasswordAuthenticationToken credentials = (UsernamePasswordAuthenticationToken) authentication;
        return authService.login(determineEmail(credentials), determinePassword(credentials));
    }

    private String determineEmail(Authentication authentication) {
        return authentication.getPrincipal() == null ? "NONE_PROVIDED" : authentication.getName();
    }

    private String determinePassword(Authentication authentication) {
        return authentication.getCredentials().toString();
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return UsernamePasswordAuthenticationToken.class.isAssignableFrom(authentication);
    }
}
