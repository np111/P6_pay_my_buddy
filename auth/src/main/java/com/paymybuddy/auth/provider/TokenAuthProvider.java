package com.paymybuddy.auth.provider;

import com.paymybuddy.auth.AuthService;
import com.paymybuddy.auth.AuthToken;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.stereotype.Component;

/**
 * An {@link AuthenticationProvider} for {@linkplain AuthToken token strings}, using our own
 * {@linkplain AuthService authentication service}.
 */
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Component
@Scope("singleton")
public class TokenAuthProvider implements AuthenticationProvider {
    private final AuthService authService;

    @Override
    public AuthToken authenticate(Authentication authentication) throws AuthenticationException {
        return authService.authenticateAuthToken(((AuthToken) authentication).getCredentials());
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return AuthToken.class.isAssignableFrom(authentication);
    }
}
