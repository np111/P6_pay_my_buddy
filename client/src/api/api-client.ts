import {default as _fetch} from 'isomorphic-unfetch';
import getConfig from 'next/config';
import {Router} from 'next/router';
import {getGlobalAuthGuard, NextPageWithAuthContext} from '../components/auth/app-with-auth';
import {AccessDeniedError, ApiException} from './api-exception';
import {ApiResponse} from './api-response';

export interface ApiFetchRequest {
    ctx?: NextPageWithAuthContext; // SSR page context
    authToken?: string | false;
    method?: 'GET' | 'POST';
    url: string;
    body?: any;
    neverCancel?: boolean;
}

export class ApiClient {
    private readonly _baseUrl: string;
    private readonly _abortSignal?: () => AbortSignal | undefined;

    constructor(baseUrl: string, abortSignal?: () => AbortSignal | undefined) {
        this._baseUrl = baseUrl;
        this._abortSignal = abortSignal;
    }

    fetch<T = any>({ctx, authToken, method, url, body, neverCancel}: ApiFetchRequest): Promise<ApiResponse<T>> {
        const headers: any = {};

        // Resolve auth token
        if (authToken !== false && authToken === undefined) {
            if (typeof window !== 'undefined') {
                authToken = getGlobalAuthGuard().token;
            } else {
                if (!ctx) {
                    throw new ApiException('ctx is missing in SSR context');
                }
                authToken = ctx.authGuard.token;
            }
        }
        if (authToken === false || authToken === undefined) {
            authToken = 'anonymous';
        }
        headers['X-Auth-Token'] = authToken;

        // Serialize body (and auto-detect method)
        if (body) {
            method = method || 'POST';
            body = JSON.stringify(body);
            headers['Content-Type'] = 'application/json;charset=utf-8';
        }
        method = method || 'GET';

        // Fetch
        const input: RequestInfo = this._baseUrl + url;
        const init: RequestInit = {method, headers, body};
        if (neverCancel !== true && this._abortSignal) {
            init.signal = this._abortSignal();
        }
        return _fetch(input, init).then((httpRes): Promise<ApiResponse<T>> => {
            const statusClass = Math.floor(httpRes.status / 100);
            if (statusClass === 2) {
                return this._parseJson(httpRes).then((result) => ({success: true, result}));
            } else if (statusClass === 4) {
                return this._parseJson(httpRes).then((error) => {
                    if (typeof error === 'object' && error && typeof error.code === 'string' && typeof error.message === 'string') {
                        if (error && error.type !== 'SERVICE') {
                            switch (error.code) {
                                case 'SERVER_EXCEPTION':
                                    throw new ApiException('Server-side exception (' + error.code + ')', error);
                                case 'ACCESS_DENIED':
                                    throw new AccessDeniedError(error);
                                default:
                                    throw new ApiException('Client-side exception (' + error.code + ')', error);
                            }
                        }
                        return {success: false, error};
                    }
                    throw new ApiException('Unsupported server response: incomplete error fields');
                });
            } else if (statusClass === 5) {
                throw new ApiException('Server exception: HTTP ' + httpRes.status);
            } else {
                throw new ApiException('Unsupported server response: HTTP ' + httpRes.status);
            }
        }).catch((e: any): ApiResponse<T> => {
            if (e && e.name === 'AbortError') {
                return {success: false, error: {type: 'CLIENT', code: '_ABORTED', message: 'Request was aborted by the client'}};
            }
            throw e;
        });
    }

    private _parseJson(httpRes: Response) {
        switch (httpRes.status) {
            default:
                return httpRes.json().catch((err) => {
                    throw new ApiException('Unsupported server response: invalid json', err);
                });
            case 204:
                return Promise.resolve(undefined);
            case 279:
                window.location.reload();
                return new Promise<any>((ignored) => {
                    // never completing promise, waiting for the page to reload
                });
        }
    }
}

export const apiClient = (() => {
    const {serverRuntimeConfig, publicRuntimeConfig} = getConfig();
    const apiUrl = serverRuntimeConfig.apiUrl || publicRuntimeConfig.apiUrl;
    let abortSignal;
    if (typeof window !== 'undefined' && typeof AbortController !== 'undefined') {
        let signal: AbortSignal;
        let controller: AbortController;
        const initController = () => {
            controller = new AbortController();
            signal = controller.signal;
        };
        const abortController = () => {
            controller.abort();
            initController();
        };
        initController();
        Router.events.on('routeChangeStart', abortController);
        abortSignal = () => signal;
    }
    return new ApiClient(apiUrl, abortSignal);
})();
