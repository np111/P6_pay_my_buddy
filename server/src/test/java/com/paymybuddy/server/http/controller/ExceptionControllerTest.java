package com.paymybuddy.server.http.controller;

import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.paymybuddy.auth.provider.CredentialsAuthProvider;
import com.paymybuddy.auth.provider.TokenAuthProvider;
import com.paymybuddy.server.mock.MockAuthGuard;
import com.paymybuddy.server.mock.TestControllerConfig;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.test.web.servlet.MockMvc;

import static com.paymybuddy.server.mock.MockMvcSnapshot.toMatchSnapshot;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;

@WebMvcTest(ExceptionController.class)
@Import(TestControllerConfig.class)
@ExtendWith(SnapshotExtension.class)
class ExceptionControllerTest {
    @MockBean
    private CredentialsAuthProvider credentialsAuthProvider;
    @MockBean
    private TokenAuthProvider tokenAuthProvider;

    @Autowired
    private MockMvc mockMvc;
    @Autowired
    private ObjectMapper objectMapper;

    @Test
    @MockAuthGuard.WithAuthToken
    void handleNoHandlerFoundException() throws Exception {
        toMatchSnapshot(
                mockMvc.perform(get("/no-handler-found")).andReturn()
        );
    }
}