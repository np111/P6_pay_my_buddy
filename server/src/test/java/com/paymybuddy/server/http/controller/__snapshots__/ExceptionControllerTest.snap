com.paymybuddy.server.http.controller.ExceptionControllerTest.handleNoHandlerFoundException=[
-
> GET /no-handler-found
< HTTP 400
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
{"type":"CLIENT","status":400,"code":"BAD_REQUEST","message":"No handler found for GET /no-handler-found","metadata":{}}
]